### PR TITLE
Refactor Updates

### PR DESCRIPTION
**Pull request by calvbore**
- Update hooks to match functionality with scafold-eth after azf refactor.

### CODE REVIEW

File: src/ContractReader.ts [hunk 1]
```diff
@@ -9,7 +9,8 @@ import { BLOCK_TIME } from "./constants";
 const DEBUG = false;
 
 export default function useContractReader(
-  contract: Contract,
+  contracts: {[index: string]: Contract},
```
Comment (ShravanSunder) [inline src/ContractReader.ts:12]: Consider using `Record<string, Contract>` as a type

File: src/ContractLoader.ts [hunk 2]
```diff
@@ -1,54 +1,136 @@
...
+  customAddresses?: {[index: string]: any},
+  hardhatContracts: {[index: string]: Contract},
+  externalContracts: {[index: string]: Contract}
+}
+
 export default function useContractLoader(
-  provider: Web3Provider,
-  address: string,
-  ABI: any,
+  providerOrSigner: Web3Provider,
```
Comment (ShravanSunder) [inline src/ContractLoader.ts:47]: i think we need the type to be `JsonRpcProvider | Web3Provider,` to be compatible with the main app

File: src/ContractLoader.ts [hunk 3]
```diff
@@ -1,54 +1,136 @@
...
+  - hardhatNetworkName - to hardcode the hardhat network of interest
+  - customAddresses: { contractName: 0xCustomAddress } to hardcode the address for a given named contract
+  - hardhatContracts: object following the hardhat deploy export format (Json with chainIds as keys, which have hardhat network names as keys, which contain arrays of contracts for each)
+  - externalContracts: object with chainIds as keys, with an array of contracts for each
 */
+
+type Config = {
+  chainId?: number,
+  hardhatNetworkName?: string,
+  customAddresses?: {[index: string]: any},
```
Comment (ShravanSunder) [inline src/ContractLoader.ts:41]: consider using `Record<string> : string` for these
and Record<string, Contract>

File: src/ContractLoader.ts [hunk 4]
```diff
@@ -1,54 +1,136 @@
...
-          signer = provider;
-        }
+    let active = true;
+
+    async function loadContracts() {
+      if (typeof providerOrSigner !== "undefined") {
+        console.log(`loading contracts`);
+        try {
+          // we need to check to see if this providerOrSigner has a signer or not
+          let signer: any;
```
Comment (ShravanSunder) [inline src/ContractLoader.ts:59]: the following is a typescript based option
```
 let signer: Signer | undefined = undefined;
          let accounts;
          if (providerOrSigner && providerOrSigner instanceof JsonRpcProvider) {
            accounts = await providerOrSigner.listAccounts();
            if (accounts && accounts.length > 0) {
              signer = providerOrSigner.getSigner();
            }
          }
if (!signer && providerOrSigner instanceof Signer) {
            signer = providerOrSigner as Signer;
          }
```

File: src/GasPrice.ts [hunk 5]
```diff
@@ -10,7 +10,7 @@ export default function useGasPrice(
   const [gasPrice, setGasPrice] = useState<number>();
 
   const loadGasPrice = async () => {
-    if(targetNetwork.gasPrice){
+    if(targetNetwork.hasOwnProperty("gasPrice")){
```
Comment (ShravanSunder) [inline src/GasPrice.ts:13]: you can use 
`if(targetNetwork?.gasPrice){`

File: src/ContractLoader.ts [hunk 6]
```diff
@@ -1,54 +1,136 @@
...
+            combinedContracts = { ...combinedContracts, ...externalContractList[_chainId].contracts };
+          }
+
+          const newContracts = Object.keys(combinedContracts).reduce((accumulator: {[index: string]: Contract}, contractName: string) => {
+            const _address =
+              config.customAddresses && Object.keys(config.customAddresses).includes(contractName)
+                ? config.customAddresses[contractName]
+                : combinedContracts[contractName].address;
+            accumulator[contractName] = new ethers.Contract(_address, combinedContracts[contractName].abi, signer);
+            return accumulator;
```
Comment (ShravanSunder) [inline src/ContractLoader.ts:120]: where is this coming from?  this doesn't seem to match the contract loader? 😵

Comment (calvbore) [reply src/ContractLoader.ts:120]: This is the new implementation of useContractLoader from [#192](https://github.com/austintgriffith/scaffold-eth/pull/192). It's more generalized and removes the need for the useCustomeContractLoader and useExternalContractLoader hooks.

**Pull request merged**